### PR TITLE
OBPIH-5854 New approval statuses in SM list pages - Approvals lists pages

### PR DESCRIPTION
**Pull request by alannadolny**

### CODE REVIEW

File: grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy
Comment (alannadolny) [inline grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy:831]: The idea of this function is to use an appropriate list based on the list page that we are currently working on:
1. If we are on the requests list:
   a) If our location supports request approval activity - add waiting for approval and approved
   b) If our location doesn't support request approval - the options should be the same as we had before
2. If we are on the outbound list:
   a) If our location supports request approval activity - add the approved status
   b) If our location doesn't support request approval - the options should be the same as we had before

File: grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy
Comment (alannadolny) [inline grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy:820]: I had to split options on the dropdown from statuses that we can have on the list page. In the frontend part, we used fetched statuses from the dropdown to display their names on the list, but when we removed the verified status we still had at least one stock movement with this status, and the error was thrown. So I split those two cases (We were trying to show the status that we didn't fetch). Regarding the validation with visible requests in "approved"/"waiting for approval" status - we have a ticket for this and I think this should be improved/resolved here

File: src/js/hooks/list-pages/outbound/useOutboundFilters.jsx
Comment (alannadolny) [inline src/js/hooks/list-pages/outbound/useOutboundFilters.jsx:43]: When we have different statuses on the dropdown, we should fetch them when:
- We are changing location (second location can not support requests approval, so this options should not be visible)
- We are changing locale (different translations)
- Changing list pages

File: src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy
Comment (kchelstowski) [inline src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy:81]: I don't know if it's a good naming here - what do you mean by "Requisition" here?

File: grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy
Comment (kchelstowski) [reply grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy:820]: if `options` are still based on what `getRequisitionStatusCodesList` returns, can't we still run into this problem, since this method can filter out some of the statuses?

File: src/js/actions/index.js
Comment (kchelstowski) [reply src/js/actions/index.js:534]: never mind - I see that it is also used in the `useOutboundFilters` hook, but anyway, I'll leave this comment here

**Review by kchelstowski — COMMENTED** [review]
A general comment is that we probably need some comments for the RequisitionStatus methods.
In one place you use `listRequisitionOptions`, in other place `listOutboundOptions`, in other place `listOutboundOptionsWhenSupportingRequestApproval`, in other place `listAll()` - it's easy to get a headache while reading that and we should make sure to at least add comments above them to know, what they are supposed to exclude etc. 
I agree the naming of them might say that, but for example - I don't know what we mean by `listRequisitionOptions` here, and what's the difference between this and `listOutboundOptions` since outbound is a requisition based SM :man_shrugging:

File: src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy
Comment (alannadolny) [reply src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy:81]: I changed the naming here, but I am still open to better propositions

File: src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy
Comment (kchelstowski) [inline src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy:83]: "we" => current location

File: src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy
Comment (kchelstowski) [inline src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy:85]: looking at this stuff, I've now started to think if we should not refactor those statuses methods to:

- create a common list method that would contain the intersection of all the methods
- for particular methods just concat it with the statuses that we are adding, e.g: `listCommon().concat([APPROVED])`
- or filter it if we want to exclude something

don't know if it should be part of this ticket, but I think it would improve the readability of those methods.

File: src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy
Comment (kchelstowski) [inline src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy:84]: I'm stlil not sure about the naming for this one, but I don't have any suggestion so I'll leave it for others maybe

File: src/groovy/org/pih/warehouse/inventory/StockMovementStatusCode.groovy
Comment (drodzewicz) [inline src/groovy/org/pih/warehouse/inventory/StockMovementStatusCode.groovy:21]: Shouldn't this be `PENDING_APPROVAL`?
I saw Kacper has implemented these enums in #4288 so lets stick to the same name

File: src/groovy/org/pih/warehouse/inventory/StockMovementStatusCode.groovy
Comment (drodzewicz) [inline src/groovy/org/pih/warehouse/inventory/StockMovementStatusCode.groovy:24]: `REJECTED` is probably missing right?

File: src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy
Comment (drodzewicz) [reply src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy:85]: Good point, I think we can even implement some kind of a `RequisitionStatusFactory` to generate a list of the statuses based on some given parameters.
But for now, unless anyone disagrees I think we can proceed with the individual lists

File: src/js/actions/index.js
Comment (drodzewicz) [inline src/js/actions/index.js:536]: let's use `params` attribute to pass parameters instead of template literals
```suggestion
    apiClient.get('/openboxes/api/stockMovements/requisitionsStatusCodes', {
        params: { isRequestsList },
    })
```

**Review by drodzewicz — CHANGES_REQUESTED** [review]

File: grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy
Comment (drodzewicz) [reply grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy]: I think I understand the problem but not fully. To me, this approach doesn't seem right but maybe I am missing some context.
Anyway, we will probably need to discuss this part further.

File: src/groovy/org/pih/warehouse/inventory/StockMovementStatusCode.groovy
Comment (alannadolny) [reply src/groovy/org/pih/warehouse/inventory/StockMovementStatusCode.groovy:24]: I don't need this status here

File: src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy
Comment (awalkowiak) [inline src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy:142]: Kacper's PR might add the same thing, watch out for conflicts.

File: grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy
Comment (awalkowiak) [inline grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy:818]: Why you can't just return `[data: options]` based on the case?

File: src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy
Comment (awalkowiak) [reply src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy:84]: Yeah this is something that we could think about a bit

**Review by awalkowiak — CHANGES_REQUESTED** [review]

File: grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy
Comment (kchelstowski) [inline grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy:824]: it'd be cleaner if you reverted the statement, so:
```groovy
// If a location doesn't support the request approval, return listOutboundOptions no matter what list is displayed
if (!hasRequestApprovalActivity) {
 return RequisitionStatus.listOutboundOptions()
}
// If supports request approval, check what type of list page it is and return appropriate statuses
if (isRequestList) {
 return RequisitionStatus.listRequestOptionsWhenSupportingRequestApproval()
}
return RequisitionStatus.listOutboundOptionsWhenSupportingRequestApproval()
```

File: src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy
Comment (kchelstowski) [reply src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy:84]: I assume we don't want to spend time thinking about naming this one? @awalkowiak ?

File: src/js/components/stock-movement/outbound/StockMovementOutboundTable.jsx
Comment (kchelstowski) [inline src/js/components/stock-movement/outbound/StockMovementOutboundTable.jsx:46]: nitpicky: I think this call should be placed in the `useOutboundListTableData` hook, but no need to change.

File: src/js/components/stock-movement/outbound/StockMovementOutboundTable.jsx
Comment (kchelstowski) [inline src/js/components/stock-movement/outbound/StockMovementOutboundTable.jsx:144]: please make sure that the label will change when you change the language - I remember having difficult times with `translate` behavior when changing the language on list pages - most of the times I was just deciding to translate what I needed on the backend.

File: grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy
Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy:819]: I might think about adding another enum `RequisitionStatusWorkflow` (??) to represent the different sub-workflows: RequestWorkflow, ApprovalWorkflow, PickPackWorkflow. Each status could have a workflow associated with it 
```
    CREATED(1, null, StatusType.SUCCESS, Workflow.REQUEST),
    EDITING(2, PENDING, StatusType.PRIMARY, Workflow.REQUEST),
    PENDING_APPROVAL(3, null, StatusType.WARNING, Workflow.APPROVAL),
    APPROVED(4, null, StatusType.SUCCESS, Workflow.APPROVAL),
    REJECTED(5, null, StatusType.DANGER, Workflow.APPROVAL),
    VERIFYING(6, PENDING, StatusType.WARNING, Workflow.REQUEST),
```
and you could pass a single workflow or a list of workflows to get the appropriate list back.
```
RequisitionStatus.listOptions(Workflow.ALL)
RequisitionStatus.listOptions(Workflow.REQUEST)
RequisitionStatus.listOptions([Workflow.REQUEST, Workflow.REQUEST_APPROVAL])
```
NOTE: I don't understand the need for listOutboundOptions. Could someone explain that to me?


Aside: This is one of the reasons I dislike the Stock Movement workflow so much. It's not just one workflow, it's anywhere from 3-5 different workflows, each with its own complex set of steps, that are being squashed into a single, barely usable flow.

File: grails-app/i18n/messages.properties
Comment (jmiranda) [inline grails-app/i18n/messages.properties:3487]: Can we stop duplicating these? I don't understand why the react app can just use the existing key
```
enum.RequisitionStatus.STATUS
```

**Review by jmiranda — CHANGES_REQUESTED** [review]

File: grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy
Comment (alannadolny) [reply grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy:824]: I prefer this version + changes in naming that @jmiranda suggested, than moving to the function higher + using if/else. So, I am leaving this part of the code in the version that you suggested, unless there is a strong opinion against it.

File: src/js/components/stock-movement/outbound/StockMovementOutboundTable.jsx
Comment (alannadolny) [reply src/js/components/stock-movement/outbound/StockMovementOutboundTable.jsx:144]: Eh, you're right, it doesn't work

File: grails-app/i18n/messages.properties
Comment (alannadolny) [reply grails-app/i18n/messages.properties:3487]: As I know we were using only those labels starting with "react" on the react side. For now, when I use our function for fetching translations, I have to pass "requisitionStatus" as an argument to fetch all of the translations that start with "react.requistionStatus...".

I am moving the translation to the backend because those translations don't work as @kchelstowski suggested in a comment. So, I am removing those duplications.

File: grails-app/domain/org/pih/warehouse/inventory/OutboundStockMovementListItem.groovy
Comment (alannadolny) [reply grails-app/domain/org/pih/warehouse/inventory/OutboundStockMovementListItem.groovy:111]: The variant here is an indicator that can be as an example: "warning", "success", etc. It is represented as a colorful dot on our list pages:
image

I had to move it here, because of the implementation that we had. This implementation doesn't allow us to have different options on the filtering dropdown, than the statuses on the stock movement list. In this ticket I have to remove "Verified" from the filtering when the location supports "request approval", but on the current list we can have "verified" status

File: grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy
Comment (alannadolny) [reply grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy:827]: moved

File: src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy
Comment (drodzewicz) [inline src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy:161]: I am wondering if we should specify what we are mapping to exactly.
Since we are mapping it to options maybe this should be called something like `mapToOption` or `toOption` and we can omit the "StatusCode" in the name since we are going to be calling it on RequisitionStatus so saying that we are mapping it to StatusCodes seems redundant, but that just my take.

To me this felts similar to what we are doing when mapping to JSON `.toJson()`

Comment (alannadolny) [reply src/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy:161]: Good idea, thanks

File: grails-app/domain/org/pih/warehouse/inventory/OutboundStockMovementListItem.groovy
Comment (drodzewicz) [inline grails-app/domain/org/pih/warehouse/inventory/OutboundStockMovementListItem.groovy:114]: Since `status` on `OutboundStockMovementListItem` is a **RequisitionStatus**, I feel like we can use the `mapStatusCodes()` _or whatever the name for it will be_, to extract the label from it.
I am not sure if it is going to work but lets try doing something like

```suggestion
                statusLabel         : status.mapStatusCodes()?.label,
```

**Review by drodzewicz — CHANGES_REQUESTED** [review]

File: grails-app/domain/org/pih/warehouse/inventory/OutboundStockMovementListItem.groovy
Comment (alannadolny) [reply grails-app/domain/org/pih/warehouse/inventory/OutboundStockMovementListItem.groovy]: Yup, it works

**Review by drodzewicz — APPROVED** [review]

File: src/js/hooks/list-pages/outbound/useOutboundFilters.jsx
Comment (jmiranda) [inline src/js/hooks/list-pages/outbound/useOutboundFilters.jsx:43]: Should pass source type

**Review by jmiranda — APPROVED** [review]
Minor changes if necessary otherwise LGTM

**Review by awalkowiak — CHANGES_REQUESTED** [review]

**Review by awalkowiak — APPROVED** [review]

**Pull request merged**
